### ISSUE TITLE
Cannot get a repo info which just created in a same thread.

### ISSUE THREAD

**Issue by LittleApple-fp16** [huggingface/huggingface_hub#1945]
### Describe the bug

when upload models to a new repository then get its info or using glob to extraction path right away
if these things are in a same thread, repository will be completely invisible to the thread unless using a new thread
![image](https://github.com/huggingface/huggingface_hub/assets/148928493/bd1b7cba-6979-47cd-b8b6-7ab4c5ebf7da)

### System info

```shell
- huggingface_hub version: 0.19.4
- Platform: Windows-10-10.0.19044-SP0
- Python version: 3.10.6
- Running in iPython ?: No
- Running in notebook ?: No
- Running in Google Colab ?: No
- Token path ?: E:\HuggingfaceHome\token
- Has saved token ?: True
- Who am I ?: LittleApple-fp16
- Configured git credential helpers: manager-core
- FastAI: N/A
- Tensorflow: N/A
- Torch: 2.1.2+cu118
- Jinja2: 3.1.2
- Graphviz: N/A
- Pydot: N/A
- Pillow: 9.5.0
- hf_transfer: N/A
- gradio: 4.7.1
- tensorboard: N/A
- numpy: 1.26.1
- pydantic: 2.4.2
- aiohttp: 3.8.6
- ENDPOINT: https://huggingface.co
- HF_HUB_CACHE: E:\HuggingfaceHome\hub
- HF_ASSETS_CACHE: E:\HuggingfaceHome\assets
- HF_TOKEN_PATH: E:\HuggingfaceHome\token
- HF_HUB_OFFLINE: False
- HF_HUB_DISABLE_TELEMETRY: False
- HF_HUB_DISABLE_PROGRESS_BARS: None
- HF_HUB_DISABLE_SYMLINKS_WARNING: False
- HF_HUB_DISABLE_EXPERIMENTAL_WARNING: False
- HF_HUB_DISABLE_IMPLICIT_TOKEN: False
- HF_HUB_ENABLE_HF_TRANSFER: False
- HF_HUB_ETAG_TIMEOUT: 10
- HF_HUB_DOWNLOAD_TIMEOUT: 10
```

**Comment by julien-c** [comment]
There might be a race condition and/or stale caching server-side. 

In the meantime (to confirm or infirm this) can you check whether adding a e.g. 1sec delay in the same thread exhibits the same issue?

**Comment by Wauplin** [comment]
Could you post a reproducible example for this issue @LittleApple-fp16. I'm not sure to understand if it's 1. a server-side caching issue because repo creation and "fs.exists" calls are made very closely or 2. a client-side caching issue because `HfFileSystem` has cached the information that the file doesn't exist before the file has been created.

To test if the problem comes from 1., a 1s delay as suggested by @julien-c would be a good option.
To test if the problem comes from 2., a good idea would be to reproduce the error but instead of `HfFileSystem.exists`, use `HfApi.file_exists(...)`. For the record, `HfFileSystem` is HF-compatible implementation of `fsspec` that is built to reproduce kinda a file system. It uses internally a cache system to avoid requesting multiple times the existence or not of some files. if you create a file using a `HfFileSystem` object, its internal cache is invalidated. But if you create the file separately without invalidating the fs cache, then you get inconsistencies. Please be aware of this limitation when using it. `HfApi` object on the other hand is never caching its results, meaning you have full control on what requests are made.


---
(small comment: `HF_TOKEN` environment variable is automatically read from your environment. This means that `HfFileSystem(token=os.environ.get("HF_TOKEN")` and `HfFileSystem()` are strictly equivalent. In both cases, your requests will be authenticated.)

**Comment by LittleApple-fp16** [comment]
> Could you post a reproducible example for this issue @LittleApple-fp16. I'm not sure to understand if it's 1. a server-side caching issue because repo creation and "fs.exists" calls are made very closely or 2. a client-side caching issue because `HfFileSystem` has cached the information that the file doesn't exist before the file has been created.你能为这个问题发布一个可复制的例子吗？我不确定是否是1。服务器端缓存问题，因为repo创建和“fs.exists”调用非常接近;或者2.客户端缓存问题，因为 `HfFileSystem` 在文件创建之前缓存了文件不存在的信息。
> 
> To test if the problem comes from 1., a 1s delay as suggested by @julien-c would be a good option.要测试问题是否来自1.，建议的1秒延迟是一个不错的选择。 To test if the problem comes from 2., a good idea would be to reproduce the error but instead of `HfFileSystem.exists`, use `HfApi.file_exists(...)`. For the record, `HfFileSystem` is HF-compatible implementation of `fsspec` that is built to reproduce kinda a file system. It uses internally a cache system to avoid requesting multiple times the existence or not of some files. if you create a file using a `HfFileSystem` object, its internal cache is invalidated. But if you create the file separately without invalidating the fs cache, then you get inconsistencies. Please be aware of this limitation when using it. `HfApi` object on the other hand is never caching its results, meaning you have full control on what requests are made. 为了测试问题是否来自2.，一个好主意是重现错误，但不是使用 `HfFileSystem.exists` ，而是使用 `HfApi.file_exists(...)` 。需要说明的是， `HfFileSystem` 是 `fsspec` 的HF兼容实现，它是为了重现一种文件系统而构建的。它在内部使用缓存系统，以避免多次请求某些文件的存在与否。如果您使用 `HfFileSystem` 对象创建文件，则其内部缓存将失效。但是如果单独创建文件而不使fs缓存无效，那么就会出现不一致。在使用它时请注意这个限制。 `HfApi` 对象另一方面从不缓存其结果，这意味着您可以完全控制发出的请求。
> 
> (small comment: `HF_TOKEN` environment variable is automatically read from your environment. This means that `HfFileSystem(token=os.environ.get("HF_TOKEN")` and `HfFileSystem()` are strictly equivalent. In both cases, your requests will be authenticated.)（小注释： `HF_TOKEN` 环境变量会自动从您的环境中读取。这意味着 `HfFileSystem(token=os.environ.get("HF_TOKEN")` 和 `HfFileSystem()` 是严格等价的。在这两种情况下，您的请求都将被验证。）

Thanks  for your reply. I've tried many possible approaches to this error, including long-time delay and breakpoint debugging, but did not work, this also verifies what I said that this seems to only occur under certain conditions in the same thread.

**Comment by Wauplin** [comment]
@LittleApple-fp16 Can you provide a reproducible example then? If delays and breakpoints do not solve this problem, I am quite confident it comes from the cache inside `HfFileSystem` object. But to confirm that I would need to reproduce it myself. The screenshot shared in the issue description doesn't show how the file is created.

**Comment by LittleApple-fp16** [comment]
@Wauplin Yeah i've tried reproduce it [here](https://github.com/LittleApple-fp16/hf_test/actions/runs/7458738886/job/20293298696). You can see the difference between `Run Test` and `Run Post Test In New Thread` steps. i just run the operation in new python file.

**Comment by LittleApple-fp16** [comment]
> Could you post a reproducible example for this issue @LittleApple-fp16. I'm not sure to understand if it's 1. a server-side caching issue because repo creation and "fs.exists" calls are made very closely or 2. a client-side caching issue because `HfFileSystem` has cached the information that the file doesn't exist before the file has been created.
> 
> To test if the problem comes from 1., a 1s delay as suggested by @julien-c would be a good option. To test if the problem comes from 2., a good idea would be to reproduce the error but instead of `HfFileSystem.exists`, use `HfApi.file_exists(...)`. For the record, `HfFileSystem` is HF-compatible implementation of `fsspec` that is built to reproduce kinda a file system. It uses internally a cache system to avoid requesting multiple times the existence or not of some files. if you create a file using a `HfFileSystem` object, its internal cache is invalidated. But if you create the file separately without invalidating the fs cache, then you get inconsistencies. Please be aware of this limitation when using it. `HfApi` object on the other hand is never caching its results, meaning you have full control on what requests are made.
> 
> (small comment: `HF_TOKEN` environment variable is automatically read from your environment. This means that `HfFileSystem(token=os.environ.get("HF_TOKEN")` and `HfFileSystem()` are strictly equivalent. In both cases, your requests will be authenticated.)

Also tried HfApi.file_exists() [here](https://github.com/LittleApple-fp16/hf_test/actions/runs/7458988956/job/20294050153)
HfApi can solve this problem, but HfFS with no token param also.

**Comment by LittleApple-fp16** [comment]
> @Wauplin Yeah i've tried reproduce it [here](https://github.com/LittleApple-fp16/hf_test/actions/runs/7458738886/job/20293298696). You can see the difference between `Run Test` and `Run Post Test In New Thread` steps. i just using same operation in new python file.

**Comment by Wauplin** [comment]
Thanks for the reproducible example @LittleApple-fp16. Something is off here, I'll investigate it.
